### DISCUSSION TITLE
How to find out what led to a DependencyMatcher result?

### DISCUSSION THREAD

**Discussion by songyang-dev** [explosion/spaCy#9844] [Help: Best practices]
Using `DependencyMatcher`, we get an answer with
```python
matches = matcher(doc)
```
The documentation says you can read which `pattern : List[Dict]` led to this match.
```py
matched_rule = pattern[i]["RIGHT_ID"]
```
where `i` is the token id from
```py
print(matches) # [(4851363122962674176, [6, 0, 10, 9])]
# Each token_id corresponds to one pattern dict
match_id, token_ids = matches[0]
for i in range(len(token_ids)):
    print(pattern[i]["RIGHT_ID"] + ":", doc[token_ids[i]].text)
```
What should I do if I have many patterns added to the matcher? The [example above](https://spacy.io/usage/rule-based-matching#dependencymatcher) assumes only one `pattern` element inside the `add` method, eg.
```py
matcher.add("my pattern", [pattern])
```
I'd like to have many `pattern` variables.
```py
matcher.add("copula", [copula_pattern, expletive_pattern, null_subject_pattern])
```
And I want to know which pattern led to a given match result.

**Comment by polm** [comment]
One option here is, you can use a pattern similar to what the EntityRuler does internally for the other matchers, and use a multi-part key. So you can have `copula|base`, `copula|expletive`, `copula|null`, and first process the results individually, and then group them by the key prefix (splitting the key on the `|`).

**Reply by songyang-dev** [reply]
you are suggesting that I call separated `add` methods with different arguments?

**Reply by polm** [reply]
Yes, just add your rules in multiple calls to `add` with different labels.
